{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f8fc6946-3f8f-487b-ac20-ccc9d58e56e1",
   "metadata": {
    "tags": []
   },
   "source": [
    "# **TITLE**\n",
    "### Author : Andrea Grecu"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23c04c8d-eea7-45fb-8086-dd44c6f07d67",
   "metadata": {},
   "source": [
    "## **Background | Pūtake**\n",
    "\n",
    "Orr et al., (2020) described a phylogenomic method to measure somatic mutations within a phenotypically mosaic plant individual and further predict the somatic mutation rate within that individual. Both the bioinformatic workflow (*aka. pipeline*) and inputted data used are open-access and found within the orginal journal publication linked below.\n",
    "\n",
    "[Orr et al., ORIGINAL!](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7126060/)\n",
    "\n",
    "The original pipeline created by Orr et al., (2020) was published within a github repository utilizing many bioinformatics tools which cumitavley written in  5+ programming languages. The suggested method to replicate this pipeline is via using the makefiles provided.\n",
    "\n",
    "### Āheitanga\n",
    "\n",
    "***While the original pipeline is open access, it is not particulary user friendly.***\n",
    "\n",
    "Computational methods within bioinformatics often fall short in reproducbility due to insufficient or inacessible documentation (Birmingham, 2017). It is imperative to the progression of bioinformatics research that an effort is made to produce interactive methods which are accessible to a wider audience with limited computer literacy. \n",
    "\n",
    "> ### Thus, the purpose of this notebook is to provide an easy to follow replication of the original pipeline created by Orr et al., (2020) to detect somatic mutations. \n",
    ">The notebook provides a *computational narrative* incorporating inputs, outputs and easy to understand explanations of each step. \n",
    "> By detailing the assumptions,logic, necessary input data and expected ouput this notebook should enable this pipeline to be applied to a new set of collected data.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4479475f-03ac-4f6a-9695-bf89e1dd7bdf",
   "metadata": {},
   "source": [
    "## **Pipeline Technological Limitations | Tepenga**\n",
    "\n",
    "Considering this pipeline utilizes many (24+) whole genome reads, it requires great computational capacity which restricts the technology which it can be succesfully run on.\n",
    "\n",
    "While some changes can be made to the scripts to adapt for different RAM and thread values of your device, it is not recommended to run this pipeline on a device with capacity much lower than the default (*as this will likely take a very long time to run and much of the memory of your device*). \n",
    "\n",
    "> ### Default Capacity\n",
    "> **RAM = 64 GB**\n",
    ">\n",
    "> **CPU = 20 Threads**\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc030401-943e-498b-b82d-8cabeceb38d6",
   "metadata": {},
   "source": [
    "## **Required Software Installation |  Tāuta**\n",
    "\n",
    "Prior to running this pipeline, the appropiate software's must be installed. \n",
    "This can be done directly within this notebook by running the cells below."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a69e977-8284-4e03-b6e3-de6c1d9b05d0",
   "metadata": {},
   "source": [
    "#### Bioconda \n",
    "Most of the packages used further along in this pipeline are provided via the **Bioconda** channel (package manager). \n",
    "\n",
    "More information about bioconda can be found [here](https://bioconda.github.io/)\n",
    "\n",
    "***To install bioconda run the code cell below.***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e8e4a15d-f2be-4be9-9d5c-0b849b566bf8",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: 'defaults' already in 'channels' list, moving to the top\n",
      "Warning: 'bioconda' already in 'channels' list, moving to the top\n",
      "Warning: 'conda-forge' already in 'channels' list, moving to the top\n"
     ]
    }
   ],
   "source": [
    "import sys \n",
    "!conda config --add channels defaults\n",
    "!conda config --add channels bioconda\n",
    "!conda config --add channels conda-forge\n",
    "!conda config --set channel_priority strict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9437073e-10b3-439b-97e6-e802cacf338d",
   "metadata": {},
   "source": [
    "#### Khmer\n",
    "\n",
    "The khmer software allows for nucleotide sequence analysis, and is necessary for step (* COME BACK) -> uses the clean reads shell (Crusoe et al., 2015). \n",
    "\n",
    "More information on Khmer can be found [here](https://github.com/dib-lab/khmer)\n",
    "\n",
    "***To install Khmer run the code cell below.***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5924bebf-68b9-4c3b-b033-50a47224db79",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 22.11.0\n",
      "  latest version: 22.11.1\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c defaults conda\n",
      "\n",
      "Or to minimize the number of packages updated during conda update use\n",
      "\n",
      "     conda install conda=22.11.1\n",
      "\n",
      "\n",
      "\n",
      "# All requested packages already installed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!conda install --yes --prefix {sys.prefix} khmer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47c2ea08-92b5-4652-a4c2-83b65ceaebac",
   "metadata": {},
   "source": [
    "#### Rcorrector\n",
    "\n",
    "The Rcorrector software allows for the correction of Illumina RNA-seq , and is necessary for cleaning the reads (STEP 1*)(Song & Florea, 2015). \n",
    "\n",
    "More information on Rcorrector can be found [here](https://github.com/mourisl/Rcorrector)\n",
    "\n",
    "***To install Rcorrector run the code cell below.***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "1258991f-4456-4d7e-b6b5-652336e19b80",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1794.94s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving notices: ...working... done\n",
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 22.11.0\n",
      "  latest version: 22.11.1\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c defaults conda\n",
      "\n",
      "Or to minimize the number of packages updated during conda update use\n",
      "\n",
      "     conda install conda=22.11.1\n",
      "\n",
      "\n",
      "\n",
      "# All requested packages already installed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!conda install --yes --prefix {sys.prefix} rcorrector"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e3749a0-6e7a-4819-8b81-cd31563d6072",
   "metadata": {},
   "source": [
    "#### NextGenMap\n",
    "\n",
    "The NextGenMap software allows for short read mapping with a high sensitivity threshold, and is necessary for step ?* (Sedlazeck et al., 2013).\n",
    "\n",
    "More information on NextGenMap can be found [here](https://github.com/Cibiv/NextGenMap/wiki)\n",
    "\n",
    "***To install NextGenMap run the code cell below.***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1841983f-b3fa-44e6-a0b0-6aba5dbeaa26",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 22.11.0\n",
      "  latest version: 22.11.1\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c defaults conda\n",
      "\n",
      "Or to minimize the number of packages updated during conda update use\n",
      "\n",
      "     conda install conda=22.11.1\n",
      "\n",
      "\n",
      "\n",
      "# All requested packages already installed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!conda install --yes --prefix {sys.prefix} nextgenmap"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40336b83-39fb-49df-97f0-89aadbb95d48",
   "metadata": {},
   "source": [
    "#### GNU Parallel\n",
    "\n",
    "The GNU Parallel operating system is a free software used to conduct jobs in parallel, and is necessary for step ?* (Tange, 2018).\n",
    "\n",
    "More information on GNU Parallel can be found [here](https://www.gnu.org/software/parallel/)\n",
    "\n",
    "***To install GNU Parallel run the code cell below.***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6c7fba73-5faa-4f73-b994-5ebbf070b744",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 22.11.0\n",
      "  latest version: 22.11.1\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c defaults conda\n",
      "\n",
      "Or to minimize the number of packages updated during conda update use\n",
      "\n",
      "     conda install conda=22.11.1\n",
      "\n",
      "\n",
      "\n",
      "# All requested packages already installed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!conda install --yes --prefix {sys.prefix} parallel"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1adc27d7-99f6-406c-a504-f01b364fcb76",
   "metadata": {},
   "source": [
    "#### Samtools\n",
    "\n",
    "The Samtools software enables manipulation of next-generation sequencing data, and is necessary for step ?* (Danecek et al., 2021).\n",
    "\n",
    "More information on Samtools can be found [here](https://github.com/samtools/samtools)\n",
    "\n",
    "***To install Samtools run the code cell below.***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2d7ea843-b4b0-4ba0-a274-bb7433663693",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 22.11.0\n",
      "  latest version: 22.11.1\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c defaults conda\n",
      "\n",
      "Or to minimize the number of packages updated during conda update use\n",
      "\n",
      "     conda install conda=22.11.1\n",
      "\n",
      "\n",
      "\n",
      "# All requested packages already installed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!conda install --yes --prefix {sys.prefix} samtools"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b06a9546-90a9-4c6a-9439-11fd0507f5c0",
   "metadata": {},
   "source": [
    "#### BCFtools\n",
    "\n",
    "The BCFtools software provides commands used by Samtools and HTSlib which are adjacently installed (Danecek et al., 2021).\n",
    "\n",
    "More information on BCFtools can be found [here](https://github.com/samtools/bcftools)\n",
    "\n",
    "***To install BFCtools run the code cell below.***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a586e825-fbdb-4d1f-9202-36caa131d678",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 22.11.0\n",
      "  latest version: 22.11.1\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c defaults conda\n",
      "\n",
      "Or to minimize the number of packages updated during conda update use\n",
      "\n",
      "     conda install conda=22.11.1\n",
      "\n",
      "\n",
      "\n",
      "# All requested packages already installed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!conda install --yes --prefix {sys.prefix} bcftools"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44e52d2e-88d2-4057-b050-17faa48ea7d1",
   "metadata": {},
   "source": [
    "#### HTSlib\n",
    "\n",
    "The HTSlib is a C library used for high-throughput sequencing formats used within Samtools and BCFtools (Bonfield et al., 2021).\n",
    "\n",
    "More information on HTSlib can be found [here](https://github.com/samtools/htslib)\n",
    "\n",
    "***To install HTSlib run the code cell below.***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "15f14fb8-64dc-433c-92c7-64625d1e2735",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 22.11.0\n",
      "  latest version: 22.11.1\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c defaults conda\n",
      "\n",
      "Or to minimize the number of packages updated during conda update use\n",
      "\n",
      "     conda install conda=22.11.1\n",
      "\n",
      "\n",
      "\n",
      "# All requested packages already installed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!conda install --yes --prefix {sys.prefix} htslib"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d93f8ff0-51c6-4ab0-be2e-65a0ddd1f151",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### GATK\n",
    "\n",
    "GATK is a **G**enome **A**nalysis **T**ool**K**it designed to identify variants in genomes and is used throughout the pipeline (O’Connor & Van der Auwera, 2020).\n",
    "\n",
    "More information on GATK can be found [here](https://gatk.broadinstitute.org/hc/en-us)\n",
    "\n",
    "***To install GATK run the code cell below.***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d87fc2d0-3f2e-41bb-817f-6e1a049fb874",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 22.11.0\n",
      "  latest version: 22.11.1\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c defaults conda\n",
      "\n",
      "Or to minimize the number of packages updated during conda update use\n",
      "\n",
      "     conda install conda=22.11.1\n",
      "\n",
      "\n",
      "\n",
      "# All requested packages already installed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!conda install --yes --prefix {sys.prefix} gatk"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2649da02-74ca-4d3d-b4ed-d55228f52d68",
   "metadata": {},
   "source": [
    "#### RAxML\n",
    "\n",
    "RAxML is a **R**andomized **A**xelerated **M**aximised **L**ikelihood algorithim enabling maximum likelihood phylogenetic tree searches and is used in step ?* (Stamatakis, 2006).\n",
    "\n",
    "More information on RAxML can be found [here](https://cme.h-its.org/exelixis/web/software/raxml/index.html)\n",
    "\n",
    "***To install RAxML run the code cell below.***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1f06b05b-61e5-49bf-af51-e9e58de17b2d",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 22.11.0\n",
      "  latest version: 22.11.1\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c defaults conda\n",
      "\n",
      "Or to minimize the number of packages updated during conda update use\n",
      "\n",
      "     conda install conda=22.11.1\n",
      "\n",
      "\n",
      "\n",
      "# All requested packages already installed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!conda install --yes --prefix {sys.prefix} raxml"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9610524-f1dd-4cd3-bdf5-18e0b3cb782a",
   "metadata": {},
   "source": [
    "#### Bedtools\n",
    "\n",
    "The bedtools software encompasses many algorithimic programmes used for genome analysis and is used in step ?* (Quinlan & Hall, 2010).\n",
    "\n",
    "More information on bedtools can be found [here](https://bedtools.readthedocs.io/en/latest/)\n",
    "\n",
    "***To install Bedtools run the code cell below.***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7c00407b-fd26-4890-9415-e7b5b4bc2fa2",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 22.11.0\n",
      "  latest version: 22.11.1\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c defaults conda\n",
      "\n",
      "Or to minimize the number of packages updated during conda update use\n",
      "\n",
      "     conda install conda=22.11.1\n",
      "\n",
      "\n",
      "\n",
      "# All requested packages already installed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!conda install --yes --prefix {sys.prefix} bedtools"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e316aa6-e49a-49c0-844e-1a769f9eecf7",
   "metadata": {},
   "source": [
    "#### UCSC LiftOver\n",
    "\n",
    "LiftOver is a tool provided within the UCSC Genome Browser used to collate genetic analyses to the same build, version or collate assemblies (Hinrichs et al., 2006).\n",
    "\n",
    "More information on LiftOver can be found [here](http://hgdownload.cse.ucsc.edu/admin/exe/)\n",
    "\n",
    "***To install LiftOver run the code cell below.***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9b0b78d2-6979-4103-9caa-ca96fa93ea66",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 22.11.0\n",
      "  latest version: 22.11.1\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c defaults conda\n",
      "\n",
      "Or to minimize the number of packages updated during conda update use\n",
      "\n",
      "     conda install conda=22.11.1\n",
      "\n",
      "\n",
      "\n",
      "# All requested packages already installed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!conda install --yes --prefix {sys.prefix} ucsc-liftover"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37ff7090-d444-4e07-9c4d-5abba54e574c",
   "metadata": {},
   "source": [
    "#### VCFtools\n",
    "\n",
    "The VCFtools program package enables operations such as filtering and categorising variants of VCF (**V**ariant **C**all **F**ormat) files and is used in step ?* (Danecek et al., 2011).\n",
    "\n",
    "More information on VCFtools can be found [here](https://vcftools.github.io/)\n",
    "\n",
    "***To install VCFtools run the code cell below.***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d412a89f-361d-495f-bf54-e32be35c712b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 22.11.0\n",
      "  latest version: 22.11.1\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c defaults conda\n",
      "\n",
      "Or to minimize the number of packages updated during conda update use\n",
      "\n",
      "     conda install conda=22.11.1\n",
      "\n",
      "\n",
      "\n",
      "# All requested packages already installed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!conda install --yes --prefix {sys.prefix} vcftools"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00e44263-618b-4b86-855c-597aa47345a4",
   "metadata": {},
   "source": [
    "## **Data Input Requirements | Tāuru Raraunga**\n",
    "\n",
    "### Data Collection \n",
    "\n",
    "Orr et al.,(2020) sampled data from a phenotypically mosaic *Eucalyptus melliodora* individual (aka. a Yellow Box tree) in which it was known that one branch expressed resistance to defoliation via the *Anoploganthus* or Christmas beetle genus. From 8 distinct branches, 3 replicate samples were taken from the leaf tip in order to sequence the full genome. \n",
    "\n",
    "Data collected to be used within this pipeline is not restrictive in the number of samples but rather the number of replicates per sample;\n",
    "\n",
    "> #### **!!! TIP !!!**\n",
    ">*There must be at least 2 replicates per sample and an equal number of replicates across all samples. Each sample should be sequenced fully to produce a genome via Illumina.\n",
    "The suffix of the each replicate for one sample should be formatted as follows :*\n",
    "**Sample1a, Sample1b, Sample1c etc.**\n",
    ">\n",
    ">*The raw data inputted should be paired sequencing reads for each sample in FASTQ format. The suffix of the pair files per replicate should be* **\"R1.fastq\"** *and* **\"R2.fastq\"**\n",
    ">\n",
    ">> Raw Data should be deposited in the raw folder (/rep_files/data/raw/), in a new folder \"my_data\". You will see the raw data used by Orr et al.,(2020) in the raw folder which will be used by default.\n",
    "Each **replicate** of one **sample** should have two files in the following format: ***\"Sample1a_R1.fastq\"*** !!!!!!\n",
    "\n",
    "As always, considerations into collecting and using data in a respectful and responsible manner towards communities (across all taxa) should take precedence. \n",
    "\n",
    "### Pseudogenome\n",
    "\n",
    "There was no high-quality reference genome available for *E. melliodora* thus a pseudoreference genome was created using the reference genome of the closely related *Eucalyptus grandis* aka. Rose Gum tree (Bartholomé et al., 2014). \n",
    "> #### **!!! TIP !!!**\n",
    ">*If replicating with your own data, use either the most closely related high quality reference genome available for your sampled taxa OR for your exact taxa if available.*\n",
    ">\n",
    ">*The genome should be in a Fa-file format within a new folder inside the data folder labelled \"my_ref\" (rep_files/data/my_ref/) and the file itself called \"ref.fa\".*\n",
    ">\n",
    ">The *E.grandis* genome found in the \"e_grandis\" folder (/rep_files/data/e_grandis/) will be used by default otherwise.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "710b40d6-548a-40f2-9a78-c138d499dfc8",
   "metadata": {},
   "source": [
    "# **Analysis Makefile**\n",
    "The first makefile that Orr et al., (2020) suggests executing is found within the analysis folder of the repository files. This makefile uses 4 different scripts from the scripts folder, and can be broken down into **? main steps**.\n",
    "> *Below are descriptions of each stage considering inputs, ouputs, set variables and any potential errors.*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "972e226c-4089-4b5f-bf54-fdb0f25b03e6",
   "metadata": {},
   "source": [
    "### **Step One | Read Correction**\n",
    "The first step of the pipeline is to *correct* the raw reads.\n",
    "\n",
    "This step utilizes the algorthims provided by the **[Rcorrector](https://www.researchgate.net/publication/283260409_Rcorrector_efficient_and_accurate_error_correction_for_Illumina_RNA-seq_reads)** software to determine trusted kmers (using a De Brujin Graph) which will be further used in the next step to correct random sequencing errors producing sliced reads.\n",
    "\n",
    "#### **Input Data**\n",
    "> Raw reads with the suffix \"R1.fastq\" within the directory specified via the **READSFOLDER** variable are utilized. *I.e. the first of the paired reads for one replicate*. By default this is **\"../data/raw/\"**.\n",
    "> \n",
    "> If using your own raw reads, you may change the directory utilising the code in the stage one cell below.\n",
    "\n",
    "#### **Scripts Used**\n",
    "> Lines 100-103 of the script clean_reads.sh execute this step. \n",
    ">\n",
    "> The clean_reads.sh script is called in line 25 of the Makefile\n",
    "> (This utilizies the directories set via **SCRIPTDIR** and **CLEANREADS**).\n",
    "\n",
    "#### **Output Produced**\n",
    "> The corrected reads will be outputted in a folder named \"corrected\"found in the \"./cleaned_reads\" folder.\n",
    "> \n",
    "> The code in the cell below can be used to check if the correct files have been  produced in the corrected folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c7416a38-0b61-488e-a356-f70351e13f08",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR9650833_RRR1.cor.fq\tSRR9650841_RRR1.cor.fq\tSRR9650849_RRR1.cor.fq\n",
      "SRR9650833_RRR2.cor.fq\tSRR9650841_RRR2.cor.fq\tSRR9650849_RRR2.cor.fq\n",
      "SRR9650834_RRR1.cor.fq\tSRR9650842_RRR1.cor.fq\tSRR9650850_RRR1.cor.fq\n",
      "SRR9650834_RRR2.cor.fq\tSRR9650842_RRR2.cor.fq\tSRR9650850_RRR2.cor.fq\n",
      "SRR9650835_RRR1.cor.fq\tSRR9650843_RRR1.cor.fq\tSRR9650851_RRR1.cor.fq\n",
      "SRR9650835_RRR2.cor.fq\tSRR9650843_RRR2.cor.fq\tSRR9650851_RRR2.cor.fq\n",
      "SRR9650836_RRR1.cor.fq\tSRR9650844_RRR1.cor.fq\tSRR9650852_RRR1.cor.fq\n",
      "SRR9650836_RRR2.cor.fq\tSRR9650844_RRR2.cor.fq\tSRR9650852_RRR2.cor.fq\n",
      "SRR9650837_RRR1.cor.fq\tSRR9650845_RRR1.cor.fq\tSRR9650853_RRR1.cor.fq\n",
      "SRR9650837_RRR2.cor.fq\tSRR9650845_RRR2.cor.fq\tSRR9650853_RRR2.cor.fq\n",
      "SRR9650838_RRR1.cor.fq\tSRR9650846_RRR1.cor.fq\tSRR9650854_RRR1.cor.fq\n",
      "SRR9650838_RRR2.cor.fq\tSRR9650846_RRR2.cor.fq\tSRR9650854_RRR2.cor.fq\n",
      "SRR9650839_RRR1.cor.fq\tSRR9650847_RRR1.cor.fq\tSRR9650855_RRR1.cor.fq\n",
      "SRR9650839_RRR2.cor.fq\tSRR9650847_RRR2.cor.fq\tSRR9650855_RRR2.cor.fq\n",
      "SRR9650840_RRR1.cor.fq\tSRR9650848_RRR1.cor.fq\tSRR9650856_RRR1.cor.fq\n",
      "SRR9650840_RRR2.cor.fq\tSRR9650848_RRR2.cor.fq\tSRR9650856_RRR2.cor.fq\n"
     ]
    }
   ],
   "source": [
    "# STEP ONE CODE\n",
    "# TIP - Remove hash from code below and run cell to execute!\n",
    "\n",
    "# Use own raw data \n",
    "#!sed -i '/READSFOLDER=/c\\READSFOLDER=*path/to/your/raw/data*' *your/path/here*/analysis/Makefile\n",
    "\n",
    "# Check Corrected Reads \n",
    "#!ls [your path]rep_files/analysis/cleaned_reads/corrected"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32130ea3-fd8b-4758-a2f6-a9c0f6b42174",
   "metadata": {},
   "source": [
    "### **Step Two | Khmer Count Graph**\n",
    "The second step is to store trusted kmers in a khmer graph, which will then be used to filter corrected reads which are excessively high coverage to reduce errors when mapping reads to the reference genome. This is done using the [Khmer](https://www.researchgate.net/publication/282249513_The_khmer_software_package_Enabling_efficient_nucleotide_sequence_analysis) software.\n",
    "\n",
    "#### **Input Data**\n",
    "> Corrected reads produced from step one which were deposited in the \"./cleaned_reads/corrected\" folder are used. \n",
    "\n",
    "#### **Scripts Used**\n",
    "> Line 106 of the script clean_reads.sh executes this step, calling on the script load-into-counting-COR.py provided (due to changes in Khmer software scripts).\n",
    ">\n",
    "> The clean_reads.sh script is called in line 25 of the Makefile\n",
    "> (This utilizies the directories set via **SCRIPTDIR** and **CLEANREADS**).\n",
    "\n",
    "#### **Outputs Produced**\n",
    "> The Khmer graph itself is ouputted within the \"./cleaned_reads\" folder in a Binary file labelled **\"khmer_count.graph\"** alongside a textfile labelled **\"khmer_count.graph.info\"**.\n",
    ">\n",
    "> The textfile displays which khmer software version is installed, which files kmers were obtained from, the total number of unique kmers obtained and the **fp** *(false positive)* **rate**.\n",
    ">\n",
    "> If the last two values are displayed as zero, it is likely that an error occured!\n",
    "> **Check the values by running the code in the cells below!**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "153ed0e8-056c-44e3-b986-51910d9ff4dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of unique k-mers: 4536784228\n",
      "fp rate estimated to be 0.004\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# STEP TWO CODE\n",
    "\n",
    "# Run the lines below to check the fp rate and total khmer value\n",
    "\n",
    "import os\n",
    "os.getcwd() #run this line first to get your own cleaned_reads folder file path and replace below\n",
    "os.chdir('[your path]/SRS-AG/rep_files/analysis/cleaned_reads')\n",
    "os.getcwd()# run to check working directory correct\n",
    "!tail -3 khmer_count.graph.info"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "732aa26a-1252-48c1-9015-6d8161addda7",
   "metadata": {
    "tags": []
   },
   "source": [
    "### **Step Three | Read Slicing**\n",
    "The third step of the pipeline is to slice the corrected reads, removing reads which have high or low coverage level. \n",
    "\n",
    "The default maximum coverage specified is 40000, however, this can be altered in the cell below. \n",
    "\n",
    "*Information on filtering reads using coverage and what threshold might be appropiate for your data can be found [here](https://khmer-recipes.readthedocs.io/en/latest/001-extract-reads-by-coverage/).*\n",
    "\n",
    "#### **Input Data**\n",
    "> Corrected reads produced from step one which were deposited in the \"./cleaned_reads/corrected\" folder are used as well as their respective unique kmers stored in the khmer_count.graph file in the same folder.\n",
    "\n",
    "#### **Scripts Used**\n",
    "> Lines 108-122 of the script clean_reads.sh executes this step, calling on the script slice-paired-reads-coverage.py provided (altered from khmer's script to handle paired-end reads). \n",
    ">\n",
    "> The clean_reads.sh script is called in line 25 of the Makefile\n",
    "> (This utilizies the directories set via **SCRIPTDIR** and **CLEANREADS**).\n",
    "\n",
    "#### **Outputs Produced**\n",
    "> Paired reads which passed the coverage filtering will be outputted in a folder \"sliced\" found in the \"./cleaned_reads\" folder. \n",
    ">\n",
    "> Reads which passed filtering while their pair did not will also be outputted in the sliced folder and can be identified with the suffix \".cor_singletons.fastq\"\n",
    ">\n",
    "> The code in the cell below can be used to check if the correct files have been  produced in the sliced folder. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebaf0715-cd84-40f5-b296-dc82d03bd0fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# STEP THREE CODE\n",
    "\n",
    "# Change the max coverage to desired value below-> ** = replace\n",
    "#!sed -i '/COVERAGE=/c\\COVERAGE=*VALUE*' *your/path/here*/clean_reads.sh \n",
    "\n",
    "# Check Sliced Reads \n",
    "#!ls [your_path]/rep_files/analysis/cleaned_reads/sliced"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e3c1361-00b1-4c59-b5db-77457d9bfb4d",
   "metadata": {},
   "source": [
    "### **Step Four | Allign Reads to Reference Genome**\n",
    "The fourth step of the pipeline is to align/map reads to the reference genome (*Eucalyptus grandis* in the original exp.) using the [NextGenMap](https://cibiv.github.io/NextGenMap/) algorithim.\n",
    "\n",
    "#### **Input Data**\n",
    "> The input data is denoted by -i in the ngm_aligner.sh script and changes accordingly to which repeat is occuring \n",
    ">> **!!! SEE ITERATIONS OF STEPS FOUR AND FIVE !!!**\n",
    "\n",
    "#### **Scripts Used**\n",
    "> The script ngm_alinger.sh is used for this step, found in the scripts folder of the repository.\n",
    ">\n",
    "> This script is first called in line 47 of the Makefile (*due to iterations*).\n",
    "\n",
    "#### **Outputs Produced**\n",
    "> The output file of step four is a bam file (*aka. a Binary Alignment Map*) of the aligned reads denoted by the variable -o in ngm_aligner.sh which by default is the reference genome file name and iteration number in the following format **\"e_mel_1.bam\"**. \n",
    ">\n",
    "> To check if the bam file per iteration has been produced and is not empty run the code in the cell below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd9d8295-994e-465d-8b30-cbbbca7b5764",
   "metadata": {},
   "outputs": [],
   "source": [
    "# STEP FOUR CODE\n",
    "\n",
    "# Change directory to folder of current iteration\n",
    "#import os\n",
    "#os.chdir('/your/path/to/analysis/e_mel_*iteration num*')\n",
    "\n",
    "# Check presence and size of bam file\n",
    "#!du -bsh *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc790a0d-0630-4315-bff9-7ec5636db762",
   "metadata": {},
   "source": [
    "### **Step Five | Create a Consensus Sequence from the BAM**\n",
    "The fifth step of the pipeline is to create a consensus FASTA sequence using the algorithims [bcftools consensus](https://samtools.github.io/bcftools/howtos/consensus-sequence.html) and [tabix](http://www.htslib.org/doc/tabix.html) both provided by samtools. \n",
    "\n",
    "The final iteration of the consensus sequence will be utilized as the pseudogenome reference genome.\n",
    "\n",
    "#### **Input Data**\n",
    "> The inputted data is the BAM file of the previous iteration produced by step 4 above.\n",
    ">> **!!! SEE ITERATIONS OF STEPS FOUR AND FIVE !!!**\n",
    "\n",
    "#### **Scripts Used**\n",
    "> The script create_consensus.sh is used for this step, found in the scripts folder of the repository.\n",
    ">\n",
    "> This script is fist called in line 35 of the Makefile (*due to iterations*).\n",
    "\n",
    "#### **Outputs Produced**\n",
    "> The consensus FASTA sequence is outputted as a file denoted by -o in create_consensus.sh which by default is theis the reference genome file name and iteration number in the following format **\"e_mel_1.fa\"**. \n",
    ">\n",
    "> Additionally, a chain file documenting each rearrangement between the input BAM file and the output consensus file. By default this has the same name as the outputted consensus sequence with the suffix \".chain\".\n",
    ">\n",
    "> While using the bcf tools algorithim a vcf file containing genotype calls is generated but by default NOT SAVED as an output file. This is because this pipeline rellies on the [GATK best practices](https://gatk.broadinstitute.org/hc/en-us/sections/360007226651)\n",
    "workflow to call the genotype variants further along in the pipeline. This vcf file was used to create the consensus sequence. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbea6fb1-799e-4b5b-9ae8-2ecaf4eb0d02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# STEP FIVE CODE\n",
    "\n",
    "# To save the bcf tools generated vcf file \n",
    "# !sed -i '/BCFTOOLSFILE=/c\\BCFTOOLSFILE=\"*your_bcf_file.vcf\"' *your-path-here*/create_consensus.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "410ed7f9-840a-458d-a374-20c117c2adaa",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/agre945/anaconda3/envs/studyenv/GitHub/SRS-AG/rep_files/analysis'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Change Directory\n",
    "import os\n",
    "os.getcwd() #run this line first \n",
    "os.chdir('/home/agre945/anaconda3/envs/studyenv/GitHub/SRS-AG/rep_files/analysis')\n",
    "os.getcwd()# run to check working directory correct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9594c08-65a0-40c5-a662-8bd320277748",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../scripts/clean_reads.sh -i ../data/raw/\n",
      "Academic tradition requires you to cite works you base your article on.\n",
      "If you use programs that use GNU Parallel to process data for an article in a\n",
      "scientific publication, please cite:\n",
      "\n",
      "  Tange, O. (2022, November 22). GNU Parallel 20221122 ('Херсо́н').\n",
      "  Zenodo. https://doi.org/10.5281/zenodo.7347980\n",
      "\n",
      "This helps funding further development; AND IT WON'T COST YOU A CENT.\n",
      "If you pay 10000 EUR you should feel free to use GNU Parallel without citing.\n",
      "\n",
      "More about funding GNU Parallel and the citation notice:\n",
      "https://www.gnu.org/software/parallel/parallel_design.html#citation-notice\n",
      "\n",
      "To silence this citation notice: run 'parallel --citation' once.\n",
      "\n",
      "min_coverage: None\n",
      "max_coverage: 40000\n",
      "... 0 0\n",
      "... 100000 185133\n",
      "... 200000 369958\n",
      "... 300000 554867\n",
      "... 400000 739583\n",
      "... 500000 924629\n",
      "... 600000 1109546\n",
      "... 700000 1294427\n",
      "... 800000 1479106\n",
      "... 900000 1663843\n",
      "... 1000000 1848460\n",
      "... 1100000 2033172\n",
      "... 1200000 2217835\n",
      "... 1300000 2402420\n",
      "... 1400000 2586994\n",
      "... 1500000 2771722\n",
      "... 1600000 2956158\n",
      "... 1700000 3140945\n",
      "... 1800000 3325880\n",
      "... 1900000 3510542\n",
      "... 2000000 3695371\n",
      "... 2100000 3879620\n",
      "... 2200000 4064362\n",
      "... 2300000 4248704\n",
      "... 2400000 4433315\n",
      "... 2500000 4617945\n",
      "... 2600000 4802383\n",
      "... 2700000 4986987\n",
      "... 2800000 5171617\n",
      "... 2900000 5355997\n",
      "... 3000000 5540687\n",
      "... 3100000 5725795\n",
      "... 3200000 5910652\n",
      "... 3300000 6095111\n",
      "... 3400000 6279772\n",
      "... 3500000 6464052\n",
      "... 3600000 6648608\n",
      "... 3700000 6833018\n",
      "... 3800000 7017283\n",
      "... 3900000 7201617\n",
      "... 4000000 7386057\n",
      "... 4100000 7570030\n",
      "... 4200000 7754286\n",
      "... 4300000 7938697\n",
      "... 4400000 8123092\n",
      "... 4500000 8307736\n",
      "... 4600000 8492268\n",
      "... 4700000 8676846\n",
      "... 4800000 8861137\n",
      "... 4900000 9045472\n",
      "... 5000000 9229867\n",
      "... 5100000 9414298\n",
      "... 5200000 9598481\n",
      "... 5300000 9782895\n",
      "... 5400000 9967414\n",
      "... 5500000 10151776\n",
      "... 5600000 10336205\n",
      "... 5700000 10520451\n",
      "... 5800000 10704829\n",
      "... 5900000 10889102\n",
      "... 6000000 11073629\n",
      "... 6100000 11257884\n",
      "... 6200000 11442140\n",
      "... 6300000 11626655\n",
      "... 6400000 11810994\n",
      "... 6500000 11995262\n",
      "... 6600000 12179819\n",
      "... 6700000 12364157\n",
      "... 6800000 12548504\n",
      "... 6900000 12733056\n",
      "... 7000000 12917770\n",
      "... 7100000 13102415\n",
      "... 7200000 13287071\n",
      "... 7300000 13471936\n",
      "... 7400000 13656702\n",
      "... 7500000 13841467\n",
      "... 7600000 14026194\n",
      "... 7700000 14210775\n",
      "... 7800000 14395458\n",
      "... 7900000 14580006\n",
      "... 8000000 14764684\n",
      "... 8100000 14949510\n",
      "... 8200000 15134110\n",
      "... 8300000 15318543\n",
      "... 8400000 15503426\n",
      "... 8500000 15688076\n",
      "... 8600000 15872967\n",
      "... 8700000 16061399\n",
      "... 8800000 16253181\n",
      "... 8900000 16444970\n",
      "... 9000000 16636853\n",
      "... 9100000 16828808\n",
      "... 9200000 17018287\n",
      "... 9300000 17203289\n",
      "... 9400000 17388086\n",
      "... 9500000 17573074\n",
      "... 9600000 17757964\n",
      "... 9700000 17942396\n",
      "... 9800000 18127260\n",
      "... 9900000 18311988\n",
      "... 10000000 18496641\n",
      "... 10100000 18681394\n",
      "... 10200000 18866002\n",
      "... 10300000 19050564\n",
      "... 10400000 19235316\n",
      "... 10500000 19419854\n",
      "... 10600000 19604548\n",
      "... 10700000 19788734\n",
      "... 10800000 19973552\n",
      "... 10900000 20158476\n",
      "... 11000000 20343120\n",
      "... 11100000 20527726\n",
      "... 11200000 20712184\n",
      "... 11300000 20896616\n",
      "... 11400000 21080939\n",
      "... 11500000 21265353\n",
      "... 11600000 21449851\n",
      "... 11700000 21634182\n",
      "... 11800000 21818559\n",
      "... 11900000 22003172\n",
      "... 12000000 22187452\n",
      "... 12100000 22371968\n",
      "... 12200000 22556684\n",
      "... 12300000 22741860\n",
      "... 12400000 22926584\n",
      "... 12500000 23110800\n",
      "... 12600000 23295286\n",
      "... 12700000 23479631\n",
      "... 12800000 23663891\n",
      "... 12900000 23848366\n",
      "... 13000000 24032678\n",
      "... 13100000 24217079\n",
      "... 13200000 24401721\n",
      "... 13300000 24586339\n",
      "... 13400000 24770671\n",
      "... 13500000 24955045\n",
      "... 13600000 25139169\n",
      "... 13700000 25323473\n",
      "... 13800000 25508014\n",
      "... 13900000 25692604\n",
      "... 14000000 25876946\n",
      "... 14100000 26061404\n",
      "... 14200000 26245652\n",
      "... 14300000 26430232\n",
      "... 14400000 26614612\n",
      "... 14500000 26798844\n",
      "... 14600000 26983229\n",
      "... 14700000 27167622\n",
      "... 14800000 27352087\n",
      "... 14900000 27536483\n",
      "... 15000000 27721048\n",
      "... 15100000 27905369\n",
      "... 15200000 28089679\n",
      "... 15300000 28274042\n",
      "... 15400000 28458312\n",
      "... 15500000 28642859\n",
      "... 15600000 28827251\n",
      "... 15700000 29012047\n",
      "... 15800000 29196645\n",
      "... 15900000 29381086\n",
      "... 16000000 29565715\n",
      "... 16100000 29750294\n",
      "... 16200000 29934962\n",
      "... 16300000 30119394\n",
      "... 16400000 30303811\n",
      "... 16500000 30488709\n",
      "... 16600000 30673647\n",
      "... 16700000 30858593\n",
      "... 16800000 31043224\n",
      "... 16900000 31227528\n",
      "... 17000000 31412102\n",
      "... 17100000 31596474\n",
      "... 17200000 31781271\n",
      "... 17300000 31965474\n",
      "... 17400000 32150410\n",
      "... 17500000 32334980\n",
      "... 17600000 32519793\n",
      "... 17700000 32704624\n",
      "... 17800000 32889916\n",
      "... 17900000 33074949\n",
      "... 18000000 33259863\n",
      "... 18100000 33445030\n",
      "... 18200000 33630189\n",
      "... 18300000 33815115\n",
      "... 18400000 34000211\n",
      "... 18500000 34184982\n",
      "... 18600000 34369581\n",
      "... 18700000 34554382\n",
      "... 18800000 34738942\n",
      "... 18900000 34923696\n",
      "... 19000000 35108129\n",
      "... 19100000 35292691\n",
      "... 19200000 35477478\n",
      "... 19300000 35662226\n",
      "... 19400000 35846720\n",
      "... 19500000 36031703\n",
      "... 19600000 36216046\n",
      "... 19700000 36400573\n",
      "... 19800000 36584999\n",
      "... 19900000 36769422\n",
      "... 20000000 36953791\n",
      "... 20100000 37138340\n",
      "... 20200000 37322859\n",
      "... 20300000 37507591\n",
      "... 20400000 37692323\n",
      "... 20500000 37876685\n",
      "... 20600000 38061244\n",
      "... 20700000 38246089\n",
      "... 20800000 38430974\n",
      "... 20900000 38615468\n",
      "... 21000000 38800136\n",
      "... 21100000 38984712\n",
      "... 21200000 39169534\n",
      "... 21300000 39353929\n",
      "... 21400000 39538249\n",
      "... 21500000 39722962\n",
      "... 21600000 39907360\n",
      "... 21700000 40091804\n",
      "... 21800000 40276284\n",
      "... 21900000 40460715\n",
      "... 22000000 40645353\n",
      "... 22100000 40829734\n",
      "... 22200000 41014374\n",
      "... 22300000 41198879\n",
      "... 22400000 41383694\n",
      "... 22500000 41568193\n",
      "... 22600000 41752803\n",
      "... 22700000 41937241\n",
      "... 22800000 42121698\n",
      "... 22900000 42306336\n",
      "... 23000000 42490942\n",
      "... 23100000 42675669\n",
      "... 23200000 42860514\n",
      "... 23300000 43045532\n",
      "... 23400000 43230303\n",
      "... 23500000 43415114\n",
      "... 23600000 43599803\n",
      "... 23700000 43784428\n",
      "... 23800000 43969116\n",
      "... 23900000 44154003\n",
      "... 24000000 44338683\n",
      "... 24100000 44523228\n",
      "... 24200000 44707950\n",
      "... 24300000 44893030\n",
      "... 24400000 45077905\n",
      "... 24500000 45263171\n",
      "... 24600000 45448263\n",
      "... 24700000 45633816\n",
      "... 24800000 45818692\n",
      "... 24900000 46004418\n",
      "... 25000000 46190422\n",
      "... 25100000 46375455\n",
      "... 25200000 46560346\n",
      "... 25300000 46745431\n",
      "... 25400000 46930550\n",
      "... 25500000 47115623\n",
      "... 25600000 47300842\n",
      "... 25700000 47486106\n",
      "... 25800000 47671624\n",
      "... 25900000 47856608\n",
      "... 26000000 48041609\n",
      "... 26100000 48226432\n",
      "... 26200000 48411303\n",
      "... 26300000 48595941\n",
      "... 26400000 48780736\n",
      "... 26500000 48965517\n",
      "... 26600000 49150229\n",
      "... 26700000 49335381\n",
      "... 26800000 49520243\n",
      "... 26900000 49705001\n",
      "... 27000000 49889909\n",
      "... 27100000 50074587\n",
      "... 27200000 50259073\n",
      "... 27300000 50443869\n",
      "... 27400000 50628505\n",
      "... 27500000 50813304\n",
      "... 27600000 50998150\n",
      "... 27700000 51182391\n",
      "... 27800000 51366858\n",
      "... 27900000 51551204\n",
      "... 28000000 51735700\n",
      "... 28100000 51920217\n",
      "... 28200000 52105205\n",
      "... 28300000 52289665\n",
      "... 28400000 52474037\n",
      "... 28500000 52658659\n",
      "... 28600000 52843389\n",
      "... 28700000 53027962\n",
      "... 28800000 53212555\n",
      "... 28900000 53397306\n",
      "... 29000000 53581720\n",
      "... 29100000 53766109\n",
      "... 29200000 53950556\n",
      "... 29300000 54134856\n",
      "... 29400000 54319458\n",
      "... 29500000 54504063\n",
      "... 29600000 54688478\n",
      "... 29700000 54872910\n",
      "... 29800000 55057218\n",
      "... 29900000 55241729\n",
      "... 30000000 55426025\n",
      "... 30100000 55610796\n",
      "... 30200000 55795196\n",
      "... 30300000 55979704\n",
      "... 30400000 56164004\n",
      "... 30500000 56348443\n",
      "... 30600000 56532715\n",
      "... 30700000 56717084\n",
      "... 30800000 56901651\n",
      "... 30900000 57085913\n",
      "... 31000000 57270518\n",
      "... 31100000 57454980\n",
      "... 31200000 57639348\n",
      "... 31300000 57824118\n",
      "... 31400000 58008856\n",
      "... 31500000 58193555\n",
      "... 31600000 58378222\n",
      "... 31700000 58562869\n",
      "... 31800000 58747662\n",
      "... 31900000 58932456\n",
      "... 32000000 59117271\n",
      "... 32100000 59302125\n",
      "... 32200000 59487095\n",
      "... 32300000 59672045\n",
      "... 32400000 59856738\n",
      "... 32500000 60041406\n",
      "... 32600000 60226199\n",
      "... 32700000 60411316\n",
      "... 32800000 60596344\n",
      "... 32900000 60781804\n",
      "... 33000000 60967160\n",
      "... 33100000 61152145\n",
      "... 33200000 61337688\n",
      "... 33300000 61522911\n",
      "... 33400000 61708772\n",
      "... 33500000 61897473\n",
      "... 33600000 62089336\n",
      "... 33700000 62281330\n",
      "... 33800000 62473305\n",
      "... 33900000 62663388\n",
      "... 34000000 62848554\n",
      "... 34100000 63033493\n",
      "... 34200000 63218633\n",
      "... 34300000 63403350\n",
      "... 34400000 63588366\n",
      "... 34500000 63773087\n",
      "... 34600000 63957970\n",
      "... 34700000 64142654\n",
      "... 34800000 64327309\n",
      "... 34900000 64512016\n",
      "... 35000000 64697020\n",
      "... 35100000 64881559\n",
      "... 35200000 65066313\n",
      "... 35300000 65251276\n",
      "... 35400000 65435808\n",
      "... 35500000 65620525\n",
      "... 35600000 65805126\n",
      "... 35700000 65989673\n",
      "... 35800000 66174315\n",
      "... 35900000 66358843\n",
      "... 36000000 66543752\n",
      "... 36100000 66728213\n",
      "... 36200000 66912581\n",
      "... 36300000 67096956\n",
      "... 36400000 67281761\n",
      "... 36500000 67466288\n",
      "... 36600000 67650807\n",
      "... 36700000 67835202\n",
      "... 36800000 68019443\n",
      "... 36900000 68203949\n",
      "... 37000000 68388273\n",
      "... 37100000 68572743\n",
      "... 37200000 68757648\n",
      "... 37300000 68941686\n",
      "... 37400000 69125983\n",
      "... 37500000 69310567\n",
      "... 37600000 69495148\n",
      "... 37700000 69679631\n",
      "... 37800000 69864035\n",
      "... 37900000 70048416\n",
      "... 38000000 70232957\n",
      "... 38100000 70417443\n",
      "... 38200000 70602075\n",
      "... 38300000 70786764\n",
      "... 38400000 70971257\n",
      "... 38500000 71155710\n",
      "... 38600000 71340424\n",
      "... 38700000 71524997\n",
      "... 38800000 71709697\n",
      "... 38900000 71894184\n",
      "... 39000000 72078851\n",
      "... 39100000 72263555\n",
      "... 39200000 72448028\n",
      "... 39300000 72632575\n",
      "... 39400000 72816797\n",
      "... 39500000 73001577\n",
      "... 39600000 73186139\n",
      "... 39700000 73370624\n",
      "... 39800000 73555394\n",
      "... 39900000 73740005\n",
      "... 40000000 73924494\n",
      "... 40100000 74108953\n",
      "... 40200000 74293643\n",
      "... 40300000 74478099\n",
      "... 40400000 74662930\n",
      "... 40500000 74847590\n",
      "... 40600000 75032328\n",
      "... 40700000 75217053\n",
      "... 40800000 75401853\n",
      "... 40900000 75586880\n",
      "... 41000000 75771790\n",
      "... 41100000 75956629\n",
      "... 41200000 76141621\n",
      "... 41300000 76326497\n",
      "... 41400000 76511715\n",
      "... 41500000 76696565\n",
      "... 41600000 76881362\n",
      "... 41700000 77066062\n",
      "... 41800000 77251086\n",
      "... 41900000 77436320\n",
      "... 42000000 77621397\n",
      "... 42100000 77806249\n",
      "... 42200000 77991360\n",
      "... 42300000 78176443\n",
      "... 42400000 78361627\n",
      "... 42500000 78547676\n",
      "... 42600000 78733894\n",
      "... 42700000 78919991\n",
      "... 42800000 79106302\n",
      "... 42900000 79293625\n",
      "... 43000000 79479457\n",
      "... 43100000 79664687\n",
      "... 43200000 79849557\n",
      "... 43300000 80034482\n",
      "... 43400000 80219583\n",
      "... 43500000 80404440\n",
      "... 43600000 80589663\n",
      "... 43700000 80774882\n",
      "... 43800000 80959898\n",
      "... 43900000 81144704\n",
      "... 44000000 81329514\n",
      "... 44100000 81514067\n",
      "... 44200000 81699154\n",
      "... 44300000 81884085\n",
      "... 44400000 82068808\n",
      "... 44500000 82253202\n",
      "... 44600000 82437775\n",
      "... 44700000 82622591\n",
      "... 44800000 82807158\n",
      "... 44900000 82991652\n",
      "... 45000000 83176087\n",
      "... 45100000 83360732\n",
      "... 45200000 83545510\n",
      "... 45300000 83729743\n",
      "... 45400000 83914373\n",
      "... 45500000 84099275\n",
      "... 45600000 84284031\n",
      "... 45700000 84468490\n",
      "... 45800000 84653195\n",
      "... 45900000 84837535\n",
      "... 46000000 85021841\n",
      "... 46100000 85206495\n",
      "... 46200000 85391346\n",
      "... 46300000 85575713\n",
      "... 46400000 85760110\n",
      "... 46500000 85944793\n",
      "... 46600000 86129432\n",
      "... 46700000 86313673\n",
      "... 46800000 86498109\n",
      "... 46900000 86682720\n",
      "... 47000000 86867359\n",
      "... 47100000 87051990\n",
      "... 47200000 87236653\n",
      "... 47300000 87420943\n",
      "... 47400000 87605344\n",
      "... 47500000 87790154\n",
      "... 47600000 87974645\n",
      "... 47700000 88159534\n",
      "... 47800000 88344134\n",
      "... 47900000 88528707\n",
      "... 48000000 88713444\n",
      "... 48100000 88898171\n",
      "... 48200000 89082553\n",
      "... 48300000 89267206\n",
      "... 48400000 89451997\n",
      "... 48500000 89636744\n",
      "... 48600000 89821786\n",
      "... 48700000 90006427\n",
      "... 48800000 90191482\n",
      "... 48900000 90376671\n",
      "... 49000000 90561652\n",
      "... 49100000 90746852\n",
      "... 49200000 90931751\n",
      "... 49300000 91116485\n",
      "... 49400000 91301253\n",
      "... 49500000 91486156\n",
      "... 49600000 91671357\n",
      "... 49700000 91856467\n",
      "... 49800000 92041572\n",
      "... 49900000 92227804\n",
      "... 50000000 92413875\n",
      "... 50100000 92599647\n",
      "... 50200000 92785865\n",
      "... 50300000 92971848\n",
      "... 50400000 93157958\n",
      "... 50500000 93344585\n",
      "... 50600000 93531502\n",
      "... 50700000 93722985\n",
      "... 50800000 93915201\n",
      "... 50900000 94107468\n",
      "... 51000000 94299814\n",
      "consumed 102033408 reads; kept 94331900\n",
      "min_coverage: None\n",
      "max_coverage: 40000\n",
      "... 0 0\n",
      "... 100000 184766\n",
      "... 200000 369450\n",
      "... 300000 553971\n",
      "... 400000 738700\n",
      "... 500000 923604\n",
      "... 600000 1108442\n",
      "... 700000 1292930\n",
      "... 800000 1477953\n",
      "... 900000 1662733\n",
      "... 1000000 1847462\n",
      "... 1100000 2032436\n",
      "... 1200000 2217435\n",
      "... 1300000 2402153\n",
      "... 1400000 2586867\n",
      "... 1500000 2771639\n",
      "... 1600000 2956254\n",
      "... 1700000 3140890\n",
      "... 1800000 3325692\n",
      "... 1900000 3510293\n",
      "... 2000000 3695288\n",
      "... 2100000 3879974\n",
      "... 2200000 4064674\n",
      "... 2300000 4249425\n",
      "... 2400000 4434097\n",
      "... 2500000 4618728\n",
      "... 2600000 4803242\n",
      "... 2700000 4987681\n",
      "... 2800000 5172189\n",
      "... 2900000 5356829\n",
      "... 3000000 5541568\n",
      "... 3100000 5726188\n",
      "... 3200000 5910849\n",
      "... 3300000 6095610\n",
      "... 3400000 6280199\n",
      "... 3500000 6464937\n",
      "... 3600000 6649788\n",
      "... 3700000 6834418\n",
      "... 3800000 7018580\n",
      "... 3900000 7203257\n",
      "... 4000000 7387665\n",
      "... 4100000 7572303\n",
      "... 4200000 7757140\n",
      "... 4300000 7941873\n",
      "... 4400000 8126457\n",
      "... 4500000 8311127\n",
      "... 4600000 8495644\n",
      "... 4700000 8680255\n",
      "... 4800000 8864921\n",
      "... 4900000 9049394\n",
      "... 5000000 9234227\n",
      "... 5100000 9419056\n",
      "... 5200000 9603870\n",
      "... 5300000 9788561\n",
      "... 5400000 9973119\n",
      "... 5500000 10157764\n",
      "... 5600000 10342296\n",
      "... 5700000 10526829\n",
      "... 5800000 10711605\n",
      "... 5900000 10896131\n",
      "... 6000000 11080846\n",
      "... 6100000 11265908\n",
      "... 6200000 11450847\n",
      "... 6300000 11635583\n",
      "... 6400000 11820139\n",
      "... 6500000 12004820\n",
      "... 6600000 12189579\n",
      "... 6700000 12374036\n",
      "... 6800000 12558742\n",
      "... 6900000 12743226\n",
      "... 7000000 12927799\n",
      "... 7100000 13112356\n",
      "... 7200000 13296894\n",
      "... 7300000 13481550\n",
      "... 7400000 13666269\n",
      "... 7500000 13850956\n",
      "... 7600000 14035412\n",
      "... 7700000 14220013\n",
      "... 7800000 14404585\n",
      "... 7900000 14589197\n",
      "... 8000000 14773999\n",
      "... 8100000 14958617\n",
      "... 8200000 15143298\n",
      "... 8300000 15328042\n",
      "... 8400000 15512688\n",
      "... 8500000 15697333\n",
      "... 8600000 15882100\n",
      "... 8700000 16066722\n",
      "... 8800000 16251513\n",
      "... 8900000 16436147\n",
      "... 9000000 16620930\n",
      "... 9100000 16806049\n",
      "... 9200000 16990901\n",
      "... 9300000 17175465\n",
      "... 9400000 17360415\n",
      "... 9500000 17545285\n",
      "... 9600000 17730158\n",
      "... 9700000 17915239\n",
      "... 9800000 18100111\n",
      "... 9900000 18285053\n",
      "... 10000000 18469963\n",
      "... 10100000 18654875\n",
      "... 10200000 18840465\n",
      "... 10300000 19025683\n",
      "... 10400000 19210459\n",
      "... 10500000 19395174\n",
      "... 10600000 19579960\n",
      "... 10700000 19764804\n",
      "... 10800000 19949308\n",
      "... 10900000 20133947\n",
      "... 11000000 20318688\n",
      "... 11100000 20503129\n",
      "... 11200000 20687942\n",
      "... 11300000 20872369\n",
      "... 11400000 21057232\n",
      "... 11500000 21241705\n",
      "... 11600000 21426373\n",
      "... 11700000 21610618\n",
      "... 11800000 21795551\n",
      "... 11900000 21980367\n",
      "... 12000000 22165205\n",
      "... 12100000 22349753\n",
      "... 12200000 22534238\n",
      "... 12300000 22719010\n",
      "... 12400000 22903990\n",
      "... 12500000 23088939\n",
      "... 12600000 23273777\n",
      "... 12700000 23458453\n",
      "... 12800000 23642883\n",
      "... 12900000 23827780\n",
      "... 13000000 24012571\n",
      "... 13100000 24197229\n",
      "... 13200000 24381821\n",
      "... 13300000 24566534\n",
      "... 13400000 24751391\n",
      "... 13500000 24936170\n",
      "... 13600000 25120838\n",
      "... 13700000 25305668\n",
      "... 13800000 25490528\n",
      "... 13900000 25675217\n",
      "... 14000000 25860081\n",
      "... 14100000 26044658\n",
      "... 14200000 26229561\n",
      "... 14300000 26414185\n",
      "... 14400000 26598922\n",
      "... 14500000 26783341\n",
      "... 14600000 26967916\n",
      "... 14700000 27152575\n",
      "... 14800000 27337273\n",
      "... 14900000 27522065\n",
      "... 15000000 27706808\n",
      "... 15100000 27891771\n",
      "... 15200000 28076005\n",
      "... 15300000 28260575\n",
      "... 15400000 28444984\n",
      "... 15500000 28629925\n",
      "... 15600000 28814606\n",
      "... 15700000 28999369\n",
      "... 15800000 29184080\n",
      "... 15900000 29368756\n",
      "... 16000000 29553232\n",
      "... 16100000 29737819\n",
      "... 16200000 29922587\n",
      "... 16300000 30107169\n",
      "... 16400000 30291873\n",
      "... 16500000 30476777\n",
      "... 16600000 30661439\n",
      "... 16700000 30845861\n",
      "... 16800000 31030955\n",
      "... 16900000 31215906\n",
      "... 17000000 31400624\n",
      "... 17100000 31585393\n",
      "... 17200000 31770185\n",
      "... 17300000 31954903\n",
      "... 17400000 32139490\n",
      "... 17500000 32324171\n",
      "... 17600000 32508838\n",
      "... 17700000 32693527\n",
      "... 17800000 32878271\n",
      "... 17900000 33062950\n",
      "... 18000000 33247680\n",
      "... 18100000 33432439\n",
      "... 18200000 33617451\n",
      "... 18300000 33802454\n",
      "... 18400000 33987150\n",
      "... 18500000 34171762\n",
      "... 18600000 34356667\n",
      "... 18700000 34541296\n",
      "... 18800000 34726096\n",
      "... 18900000 34910487\n",
      "... 19000000 35095132\n",
      "... 19100000 35279712\n",
      "... 19200000 35464556\n",
      "... 19300000 35649138\n",
      "... 19400000 35833842\n",
      "... 19500000 36018330\n",
      "... 19600000 36203069\n",
      "... 19700000 36387865\n",
      "... 19800000 36572900\n",
      "... 19900000 36757928\n",
      "... 20000000 36942766\n",
      "... 20100000 37127567\n",
      "... 20200000 37312098\n",
      "... 20300000 37496712\n",
      "... 20400000 37681375\n",
      "... 20500000 37865762\n",
      "... 20600000 38050290\n",
      "... 20700000 38235130\n",
      "... 20800000 38419722\n",
      "... 20900000 38604107\n",
      "... 21000000 38788792\n",
      "... 21100000 38973407\n",
      "... 21200000 39158072\n",
      "... 21300000 39342762\n",
      "... 21400000 39527229\n",
      "... 21500000 39711667\n",
      "... 21600000 39896225\n",
      "... 21700000 40080765\n",
      "... 21800000 40265278\n",
      "... 21900000 40449713\n",
      "... 22000000 40634480\n",
      "... 22100000 40819313\n",
      "... 22200000 41003711\n",
      "... 22300000 41188428\n",
      "... 22400000 41373111\n",
      "... 22500000 41557933\n",
      "... 22600000 41742777\n",
      "... 22700000 41927739\n",
      "... 22800000 42112322\n",
      "... 22900000 42296870\n",
      "... 23000000 42481472\n",
      "... 23100000 42666076\n",
      "... 23200000 42850546\n",
      "... 23300000 43035388\n",
      "... 23400000 43219824\n",
      "... 23500000 43404583\n",
      "... 23600000 43589280\n",
      "... 23700000 43773935\n",
      "... 23800000 43958770\n",
      "... 23900000 44143146\n",
      "... 24000000 44327922\n",
      "... 24100000 44512571\n",
      "... 24200000 44696952\n",
      "... 24300000 44881807\n",
      "... 24400000 45066583\n",
      "... 24500000 45250965\n",
      "... 24600000 45435434\n",
      "... 24700000 45620270\n",
      "... 24800000 45804878\n",
      "... 24900000 45989625\n",
      "... 25000000 46174420\n",
      "... 25100000 46359174\n",
      "... 25200000 46544281\n",
      "... 25300000 46728972\n",
      "... 25400000 46913465\n",
      "... 25500000 47098036\n",
      "... 25600000 47282897\n",
      "... 25700000 47467518\n",
      "... 25800000 47652140\n",
      "... 25900000 47836885\n",
      "... 26000000 48021529\n",
      "... 26100000 48206264\n",
      "... 26200000 48390808\n",
      "... 26300000 48575507\n",
      "... 26400000 48760223\n",
      "... 26500000 48944948\n",
      "... 26600000 49129849\n",
      "... 26700000 49314442\n",
      "... 26800000 49499048\n",
      "... 26900000 49683740\n",
      "... 27000000 49868357\n",
      "... 27100000 50052840\n",
      "... 27200000 50237683\n",
      "... 27300000 50422562\n",
      "... 27400000 50607256\n",
      "... 27500000 50792246\n",
      "... 27600000 50977013\n",
      "... 27700000 51162115\n",
      "... 27800000 51347084\n",
      "... 27900000 51532136\n",
      "... 28000000 51717315\n",
      "... 28100000 51902424\n",
      "... 28200000 52087238\n",
      "... 28300000 52271996\n",
      "... 28400000 52457088\n",
      "... 28500000 52641950\n",
      "... 28600000 52826813\n",
      "... 28700000 53012194\n",
      "... 28800000 53197030\n",
      "... 28900000 53382243\n",
      "... 29000000 53567509\n",
      "... 29100000 53752491\n",
      "... 29200000 53937919\n",
      "... 29300000 54126236\n",
      "... 29400000 54318738\n",
      "... 29500000 54511142\n",
      "... 29600000 54703433\n",
      "... 29700000 54890462\n",
      "... 29800000 55075292\n",
      "... 29900000 55260203\n",
      "... 30000000 55444727\n",
      "... 30100000 55629536\n",
      "... 30200000 55814300\n",
      "... 30300000 55998884\n",
      "... 30400000 56183447\n",
      "... 30500000 56368145\n",
      "... 30600000 56553039\n",
      "... 30700000 56737786\n",
      "... 30800000 56922423\n",
      "... 30900000 57107242\n",
      "... 31000000 57291904\n",
      "... 31100000 57476419\n",
      "... 31200000 57661215\n",
      "... 31300000 57846049\n",
      "... 31400000 58030672\n",
      "... 31500000 58215206\n",
      "... 31600000 58399761\n",
      "... 31700000 58584314\n",
      "... 31800000 58769083\n",
      "... 31900000 58953831\n",
      "... 32000000 59138468\n",
      "... 32100000 59323131\n",
      "... 32200000 59508148\n",
      "... 32300000 59692846\n",
      "... 32400000 59877706\n",
      "... 32500000 60062463\n",
      "... 32600000 60247184\n",
      "... 32700000 60432135\n",
      "... 32800000 60617012\n",
      "... 32900000 60801683\n",
      "... 33000000 60986202\n",
      "... 33100000 61170982\n",
      "... 33200000 61355824\n",
      "... 33300000 61540595\n",
      "... 33400000 61725335\n",
      "... 33500000 61910095\n",
      "... 33600000 62094753\n",
      "... 33700000 62279639\n",
      "... 33800000 62463931\n",
      "... 33900000 62648391\n",
      "... 34000000 62832943\n",
      "... 34100000 63017488\n",
      "... 34200000 63202458\n",
      "... 34300000 63387139\n",
      "... 34400000 63571851\n",
      "... 34500000 63756314\n",
      "... 34600000 63941028\n",
      "... 34700000 64125706\n",
      "... 34800000 64310500\n",
      "... 34900000 64495253\n",
      "... 35000000 64680049\n",
      "... 35100000 64864821\n",
      "... 35200000 65049286\n",
      "... 35300000 65234035\n",
      "... 35400000 65419035\n",
      "... 35500000 65603447\n",
      "... 35600000 65787909\n",
      "... 35700000 65972640\n",
      "... 35800000 66157613\n",
      "... 35900000 66342243\n",
      "... 36000000 66527048\n",
      "... 36100000 66712168\n",
      "... 36200000 66897051\n",
      "... 36300000 67081745\n",
      "... 36400000 67266822\n",
      "... 36500000 67451981\n",
      "... 36600000 67637091\n",
      "... 36700000 67822030\n",
      "... 36800000 68007213\n",
      "... 36900000 68192200\n",
      "... 37000000 68377406\n",
      "... 37100000 68562376\n",
      "... 37200000 68747188\n",
      "... 37300000 68932127\n",
      "... 37400000 69117197\n",
      "... 37500000 69302445\n",
      "... 37600000 69487466\n",
      "... 37700000 69672454\n",
      "... 37800000 69857387\n",
      "... 37900000 70042343\n",
      "... 38000000 70227012\n",
      "... 38100000 70412068\n",
      "... 38200000 70597069\n",
      "... 38300000 70782371\n",
      "... 38400000 70967558\n",
      "... 38500000 71152592\n",
      "... 38600000 71337474\n",
      "... 38700000 71522630\n",
      "... 38800000 71709648\n",
      "... 38900000 71901907\n",
      "... 39000000 72094211\n",
      "... 39100000 72286436\n",
      "... 39200000 72478805\n",
      "... 39300000 72670444\n",
      "... 39400000 72855384\n",
      "... 39500000 73040036\n",
      "... 39600000 73224863\n",
      "... 39700000 73409443\n",
      "... 39800000 73593992\n",
      "... 39900000 73778831\n",
      "... 40000000 73963563\n",
      "... 40100000 74148660\n",
      "... 40200000 74333443\n",
      "... 40300000 74518149\n",
      "... 40400000 74703114\n",
      "... 40500000 74888043\n",
      "... 40600000 75072857\n",
      "... 40700000 75257341\n",
      "... 40800000 75441885\n",
      "... 40900000 75626435\n",
      "... 41000000 75810946\n",
      "... 41100000 75995896\n",
      "... 41200000 76180870\n",
      "... 41300000 76365450\n",
      "... 41400000 76550044\n",
      "... 41500000 76734555\n",
      "... 41600000 76919377\n",
      "... 41700000 77104288\n",
      "... 41800000 77288966\n",
      "... 41900000 77473655\n",
      "... 42000000 77658396\n",
      "... 42100000 77842963\n",
      "... 42200000 78027420\n",
      "... 42300000 78212263\n",
      "... 42400000 78396986\n",
      "... 42500000 78581765\n",
      "... 42600000 78766426\n",
      "... 42700000 78951122\n",
      "... 42800000 79135912\n",
      "... 42900000 79320625\n",
      "... 43000000 79505393\n",
      "... 43100000 79690129\n",
      "... 43200000 79874866\n",
      "... 43300000 80059723\n",
      "... 43400000 80244372\n",
      "... 43500000 80428868\n",
      "... 43600000 80613711\n",
      "... 43700000 80798229\n",
      "... 43800000 80983356\n",
      "... 43900000 81168194\n",
      "... 44000000 81352836\n",
      "... 44100000 81537632\n",
      "... 44200000 81722287\n",
      "... 44300000 81907086\n",
      "... 44400000 82091776\n",
      "... 44500000 82276502\n",
      "... 44600000 82461441\n",
      "... 44700000 82646250\n",
      "... 44800000 82831002\n",
      "... 44900000 83015729\n",
      "... 45000000 83200574\n",
      "... 45100000 83385540\n",
      "... 45200000 83570093\n",
      "... 45300000 83754847\n",
      "... 45400000 83939563\n",
      "... 45500000 84124219\n",
      "... 45600000 84308889\n",
      "... 45700000 84493647\n",
      "... 45800000 84678479\n",
      "... 45900000 84863053\n",
      "... 46000000 85048076\n",
      "... 46100000 85233167\n",
      "... 46200000 85418433\n",
      "... 46300000 85603431\n",
      "... 46400000 85788862\n",
      "... 46500000 85974178\n",
      "... 46600000 86159365\n",
      "... 46700000 86344303\n",
      "... 46800000 86529222\n",
      "... 46900000 86714061\n",
      "... 47000000 86898937\n",
      "... 47100000 87083944\n",
      "... 47200000 87269191\n",
      "... 47300000 87454927\n",
      "... 47400000 87639905\n",
      "... 47500000 87825360\n",
      "... 47600000 88011011\n",
      "... 47700000 88196173\n",
      "... 47800000 88380960\n",
      "... 47900000 88565960\n",
      "... 48000000 88751161\n",
      "... 48100000 88936434\n",
      "... 48200000 89122380\n",
      "... 48300000 89307734\n",
      "... 48400000 89492970\n",
      "... 48500000 89678350\n",
      "... 48600000 89863077\n",
      "... 48700000 90048223\n",
      "... 48800000 90233999\n",
      "... 48900000 90418861\n",
      "... 49000000 90603708\n",
      "... 49100000 90788275\n",
      "... 49200000 90972992\n",
      "... 49300000 91157460\n",
      "... 49400000 91342166\n",
      "... 49500000 91526659\n",
      "... 49600000 91711203\n",
      "... 49700000 91895853\n",
      "... 49800000 92080501\n",
      "... 49900000 92264861\n",
      "... 50000000 92449606\n",
      "... 50100000 92634092\n",
      "... 50200000 92818669\n",
      "... 50300000 93003511\n",
      "... 50400000 93188486\n",
      "... 50500000 93372977\n",
      "... 50600000 93557471\n",
      "... 50700000 93741998\n",
      "... 50800000 93926560\n",
      "... 50900000 94111170\n",
      "... 51000000 94295434\n",
      "... 51100000 94480063\n",
      "... 51200000 94664566\n",
      "... 51300000 94849200\n",
      "... 51400000 95033884\n",
      "... 51500000 95218585\n",
      "... 51600000 95403174\n",
      "... 51700000 95587813\n",
      "... 51800000 95772549\n",
      "... 51900000 95957020\n",
      "... 52000000 96141781\n",
      "... 52100000 96326238\n",
      "... 52200000 96510717\n",
      "... 52300000 96695513\n",
      "... 52400000 96880127\n",
      "... 52500000 97064780\n",
      "... 52600000 97249328\n",
      "... 52700000 97434161\n",
      "... 52800000 97618691\n",
      "... 52900000 97803573\n",
      "... 53000000 97988183\n",
      "... 53100000 98172911\n",
      "... 53200000 98357818\n",
      "... 53300000 98542106\n",
      "... 53400000 98726777\n",
      "... 53500000 98911407\n",
      "... 53600000 99095896\n",
      "... 53700000 99280186\n",
      "... 53800000 99464490\n",
      "... 53900000 99649167\n",
      "... 54000000 99833686\n",
      "... 54100000 100018476\n",
      "... 54200000 100203130\n",
      "... 54300000 100387937\n",
      "... 54400000 100572485\n",
      "... 54500000 100757093\n",
      "... 54600000 100941882\n",
      "... 54700000 101126599\n",
      "... 54800000 101311101\n",
      "... 54900000 101496142\n",
      "... 55000000 101680836\n",
      "... 55100000 101865779\n",
      "... 55200000 102050577\n",
      "... 55300000 102235364\n",
      "... 55400000 102420320\n",
      "... 55500000 102605490\n",
      "... 55600000 102790290\n",
      "... 55700000 102975037\n",
      "... 55800000 103159879\n",
      "... 55900000 103344636\n",
      "... 56000000 103529433\n",
      "... 56100000 103714194\n",
      "... 56200000 103899046\n",
      "... 56300000 104083997\n",
      "... 56400000 104268766\n",
      "... 56500000 104453862\n",
      "... 56600000 104638671\n",
      "... 56700000 104824163\n",
      "... 56800000 105009177\n",
      "... 56900000 105194320\n",
      "... 57000000 105379675\n",
      "... 57100000 105564642\n",
      "... 57200000 105749780\n",
      "... 57300000 105934841\n",
      "... 57400000 106119818\n",
      "... 57500000 106305049\n",
      "... 57600000 106490467\n",
      "... 57700000 106676347\n",
      "... 57800000 106862297\n",
      "... 57900000 107048051\n",
      "... 58000000 107234164\n",
      "... 58100000 107420608\n",
      "... 58200000 107611693\n",
      "... 58300000 107802701\n",
      "... 58400000 107993639\n",
      "... 58500000 108184719\n",
      "consumed 117106412 reads; kept 108286524\n"
     ]
    }
   ],
   "source": [
    "# Export the path for the scripts used FIRST\n",
    "!export PATH=$PATH:/home/agre945/anaconda3/envs/studyenv/GitHub/SRS-AG/rep_files/scripts\n",
    "# Run line below SECOND ** = replace\n",
    "#!sed -i '/LOAD_COUNTING=/c\\LOAD_COUNTING=\"*your-path-here*/load-into-counting-COR.py\"' *your-path-here*/clean_reads.sh\n",
    "#!sed -i '/SLICE_BY_COV=/c\\SLICE_BY_COV=\"*your-path-here*/slice-paired-reads-by-coverage.py\"' *your-path-here*/clean_reads.sh\n",
    "!make"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "studyenv",
   "language": "python",
   "name": "studyenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
